### ISSUE TITLE
 Regression bug in ASAN: since Clang-7, it cannot report certain use-after-scope bugs

### ISSUE THREAD

**Issue by shao-hua-li** [llvm/llvm-project#61787]
For the following code, ASAN at -O0 failed to report the use-after-scope since Clang-7.
GCC's ASAN can detect it normally.

Compiler explorer: https://godbolt.org/z/1PEaocv9o

```shell
% cat a.c
struct a {
  signed b;
} d;
int *c;
struct a e() {
  {
    struct a f;
    c = &f.b;
    for (; 0;)
      return f;
  }
  *c = 0;
  return d;
}
int main() { e(); }
%
% clang-tk -O0 -fsanitize=address a.c && ./a.out
% 
% clang-6 -O0 -fsanitize=address a.c && ./a.out
=================================================================
==1==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7ffec29573b0 at pc 0x000000511cf8 bp 0x7ffec2957370 sp 0x7ffec2957368
WRITE of size 4 at 0x7ffec29573b0 thread T0
    #0 0x511cf7 in e /a.c:12:6
    #1 0x511e85 in main /a.c:15:14
    #2 0x7f4c02729082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082)
    #3 0x41971d in _start (/app/output.s+0x41971d)
%
```

**Comment by vitalybuka** [comment]
Looks like "for (; 0;)      return f;" prevents compiler from inserting llvm.lifetime markers.

**Comment by shao-hua-li** [comment]
Thanks for the reply. 
It is a **false negative issue**, meaning that ASAN will miss certain real bugs. I've reported quite a few similar bugs in ASAN and UBSAN. Such as 
https://github.com/llvm/llvm-project/issues/61280
https://github.com/llvm/llvm-project/issues/54667
https://github.com/llvm/llvm-project/issues/60237
